### PR TITLE
[`FSMT`] Make it compatible with `xxxForConditionalGeneration` models

### LINKED ISSUE #20824: FSTM compatibility issues with other `ForConditionalGeneration` models
### Description

The implementation of FSMT models like [`facebook/wmt19-en-de`](https://huggingface.co/facebook/wmt19-en-de) is atypical with respect to different aspects that are normally supported in other `ForConditionalGeneration` models. In particular:

- Both `FSMTModel` and `FSMTForConditionalGeneration` lack utility methods like `get_encoder` and `get_decoder`
- `FSMTForConditionalGeneration` and all its subclasses do not accept `inputs_embeds` and `decoder_inputs_embeds` as possible alternatives to `input_ids` and `decoder_input_ids` for the forward pass.

In particular, this renders the model unusable when using feature attribution methods through the `inseq` library (see related issue inseq-team/inseq#153).

### Expected behavior

We would expect all models belonging to the same family to expose a consistent API for external usage.

### System Info

- `transformers` version: 4.24.0
- Platform: Linux-5.15.0-56-generic-x86_64-with-glibc2.29
- Python version: 3.8.10
- Huggingface_hub version: 0.11.1
- PyTorch version (GPU?): 1.13.0+cu117 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: NA
- Using distributed or parallel set-up in script?: NA

### Who can help?

@ArthurZucker @younesbelkada

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/20824

`FSMT` model is an encoder-decoder model. Most of `xxxForConditionalGeneration` models can use `decoder_inputs_embeds` and `inputs_embeds` in replacement respectively of `input_ids` and `decoder_input_ids`. 

`FSMT` does not implement this functionality which breaks some assumptions made by external libraries / APIs, an issue has been flagged and reported in https://github.com/inseq-team/inseq/issues/153 

This PR fixes this behavior, by adding `inputs_embeds` and `decoder_inputs_embed` support for `FSMT` to make it consistent with other `xxxForConditionalGeneration` models. The PR also adds `get_encoder` and `get_decoder` class functions for `FSMT`, following the implementation of `T5`

`self.embed_positions(input_ids)` and `self.embed_positions(inputs_embeds[:, :, 0])` are equivalent as the positional embedding is computed with respect to the hidden states shape. But I added an extra check by assuming that 0 hidden states correspond to padding tokens, as this information is needed later by the position embedding layer. 

All FSMT slow tests pass

### CODE REVIEW

**Comment by younesbelkada** [comment]
I think that `inputs_embed` cannot be added since [positional embeddings needs to be computed too](https://github.com/huggingface/transformers/blob/ecd7de3dff7ea5713004b2f05e3869c24b8eb6e2/src/transformers/models/fsmt/modeling_fsmt.py#L721), and this requires having `input_ids`, except if the user can pass `inputs_position_embeds`  too

**Comment by younesbelkada** [comment]
Hi @gsarti 
Here is an attempt to fix #20824 , could you double check if this fixes your root issue
I ran: 
```
import inseq

model = inseq.load_model("facebook/wmt19-en-de", "integrated_gradients")
out = model.attribute(
  "The developer argued with the designer because her idea cannot be implemented.",
  n_steps=100
)
out.show()
```
but getting an error which is different from the one described in https://github.com/inseq-team/inseq/issues/153

**Comment by gsarti** [comment]
The error you see is a problem of Inseq due to force-setting `max_new_tokens` for generation, you can bypass it by passing a custom `generated_text` argument to `model.attribute` so that no call to model.generate is performed, but only forwards for the attribution.

Trying this code now:

```python
import inseq

model = inseq.load_model("facebook/wmt19-en-de", "integrated_gradients")
out = model.attribute(
  "The developer argued with the designer because her idea cannot be implemented.",
  "Hallo Welt ich bin Gabriele",
  n_steps=100
)
out.show()
```

I still encounter an error:

```shell
/usr/local/lib/python3.8/dist-packages/transformers/models/fsmt/modeling_fsmt.py in forward(self, input_ids, attention_mask, decoder_input_ids, decoder_attention_mask, head_mask, decoder_head_mask, cross_attn_head_mask, encoder_outputs, past_key_values, use_cache, output_attentions, output_hidden_states, inputs_embeds, inputs_position_embeds, decoder_inputs_embeds, decoder_inputs_position_embeds, return_dict)
   1090             decoder_padding_mask, causal_mask = None, None
   1091 
-> 1092         assert decoder_input_ids is not None
   1093 
   1094         if encoder_outputs is None:

AssertionError:
```

@younesbelkada  Is this check needed now? I presume extra tests for the forward call using embeddings as inputs would fail, too, if they were present!

**Comment by younesbelkada** [comment]
Thank you very much @gsarti for double checking, I adapted the assert condition based on your suggestion, getting now an error since `inputs_position_embeds` needs to be passed too.

**Comment by gsarti** [comment]
Is it standard to have `inputs_position_embeds` as forward inputs? Don't recall seeing them in other models. In principle if the positions are just created from sinusoidals they could be omitted and added in the forward itself, right?

File: src/transformers/models/fsmt/modeling_fsmt.py [hunk 1]
```diff
@@ -480,6 +481,8 @@ def forward(
             input_ids (`torch.LongTensor`): tokens in the source language of shape
                 *(batch, src_len)*
             attention_mask (`torch.LongTensor`): indicating which indices are padding tokens
+            inputs_embeds (`torch.FloatTensor`):
+                embedding vectors of shape *(batch, src_len, embed_dim)*
```
Comment (ArthurZucker) [inline src/transformers/models/fsmt/modeling_fsmt.py:485]:
```suggestion
            input_ids (`torch.LongTensor` of shape `(batch_size, sequence_length)`):
                Indices of input sequence tokens in the vocabulary of the source language. Padding will be ignored by default should you
                provide it.

                Indices can be obtained using [`OPTTokenizer`]. See [`PreTrainedTokenizer.encode`] and
                [`PreTrainedTokenizer.__call__`] for details.

                [What are input IDs?](../glossary#input-ids)
            attention_mask (`torch.Tensor` of shape `(batch_size, sequence_length)`, *optional*):
                Mask to avoid performing attention on padding token indices. Mask values selected in `[0, 1]`:

                - 1 for tokens that are **not masked**,
                - 0 for tokens that are **masked**.

            inputs_embeds (`torch.FloatTensor` of shape `(batch_size, sequence_length, hidden_size)`, *optional*):
                Optionally, instead of passing `input_ids` you can choose to directly pass an embedded representation.
                This is useful if you want more control over how to convert `input_ids` indices into associated vectors
                than the model's internal embedding lookup matrix.
```

File: src/transformers/models/fsmt/modeling_fsmt.py [hunk 2]
```diff
@@ -470,6 +470,7 @@ def forward(
         self,
         input_ids: torch.Tensor,
         attention_mask: Optional[torch.Tensor] = None,
+        inputs_embeds: torch.Tensor = None,
```
Comment (ArthurZucker) [inline src/transformers/models/fsmt/modeling_fsmt.py:473]:
```suggestion
        attention_mask: Optional[torch.Tensor] = None,
        inputs_embeds: Optional[torch.Tensor] = None,
```
As I believe the input embedings are optionally passed down here

File: src/transformers/models/fsmt/modeling_fsmt.py [hunk 3]
```diff
@@ -675,6 +694,7 @@ def forward(
         decoder_padding_mask: torch.Tensor,
         decoder_causal_mask: torch.Tensor,
         head_mask: Optional[torch.Tensor] = None,
+        inputs_embeds: Optional[torch.Tensor] = None,
```
Comment (ArthurZucker) [inline src/transformers/models/fsmt/modeling_fsmt.py:697]: Doctring is missing here

File: src/transformers/models/fsmt/modeling_fsmt.py [hunk 4]
```diff
@@ -1148,6 +1191,8 @@ def forward(
         cross_attn_head_mask: Optional[torch.Tensor] = None,
         encoder_outputs: Optional[Tuple[torch.FloatTensor]] = None,
         past_key_values: Optional[Tuple[torch.FloatTensor]] = None,
+        inputs_embeds: Optional[torch.Tensor] = None,
+        decoder_inputs_embeds: Optional[torch.Tensor] = None,
```
Comment (ArthurZucker) [inline src/transformers/models/fsmt/modeling_fsmt.py:1195]: These two arguments should be added to the docstring.

**Comment by ArthurZucker** [comment]
You can refer to other models for the usual docstring that we write for `input_ids` etc. Let's keep consistency wherever we can

**Review by sgugger — COMMENTED** [review]
@stas00 since you added this model, would also be interested to have your review on the PR.

**Comment by stas00** [comment]
I will try to find time to review tomorrow, but I also wanted to point out @patil-suraj's sync https://github.com/huggingface/transformers/pull/11218 which never got merged, but might be a useful reference as I think the same work was done there.

File: src/transformers/models/fsmt/modeling_fsmt.py [hunk 5]
```diff
@@ -717,15 +737,27 @@ def forward(
...
-            # assert input_ids.ne(self.padding_idx).any()
+        if input_ids is not None and inputs_embeds is not None:
+            raise ValueError("You cannot specify both decoder_input_ids and decoder_inputs_embeds at the same time")
+        elif input_ids is not None:
+            # embed positions
+            positions = self.embed_positions(input_ids)
+            if use_cache:
+                input_ids = input_ids[:, -1:]
+                positions = positions[:, -1:]  # happens after we embed them
+                # assert input_ids.ne(self.padding_idx).any()
```
Comment (stas00) [inline src/transformers/models/fsmt/modeling_fsmt.py:748]: remove debug statement?

**Review by stas00 — APPROVED** [review]
Thank you for extending this model, @younesbelkada!

**Review by sgugger — APPROVED** [review]
Good to go once comments are resolved.

File: src/transformers/models/fsmt/modeling_fsmt.py (hunk 5 above)
Comment (younesbelkada) [reply src/transformers/models/fsmt/modeling_fsmt.py]: fixed in 782df0b

File: src/transformers/models/fsmt/modeling_fsmt.py (hunk 4 above)
Comment (younesbelkada) [reply src/transformers/models/fsmt/modeling_fsmt.py:1206]: Fixed in [d3800fd](https://github.com/huggingface/transformers/pull/20825/commits/d3800fdee2cbe30440028ca0932fe1529f4b895b)

File: src/transformers/models/fsmt/modeling_fsmt.py (hunk 3 above)
Comment (younesbelkada) [reply src/transformers/models/fsmt/modeling_fsmt.py:709]: Fixed in [d3800fd](https://github.com/huggingface/transformers/pull/20825/commits/d3800fdee2cbe30440028ca0932fe1529f4b895b)
